src/providers/baremetal.rs

```diff
@@ -0,0 +1,221 @@
+use crate::ssh;
+use failure::Error;
+use std::collections::HashMap;
+use std::sync::Arc;
+
+/// Descriptor for a single, existing machine to connect to.
+/// Therefore, the `impl MachineSetup` includes the address of the machine in `region`; i.e.,
+/// each instance of Setup corresponds to a single machine.
+#[derive(Clone)]
+pub struct Setup {
+    addr: std::net::SocketAddr,
+    username: String,
+    key_path: Option<std::path::PathBuf>,
+    setup_fn:
+        Option<Arc<dyn Fn(&mut ssh::Session, &slog::Logger) -> Result<(), Error> + Send + Sync>>,
+}
+
+impl super::MachineSetup for Setup {
+    type Region = String;
+    fn region(&self) -> Self::Region {
+        format!("bare:{}", self.addr)
+    }
+}
+
+impl PartialEq for Setup {
+    fn eq(&self, other: &Self) -> bool {
+        self.addr == other.addr
+    }
+}
+
+impl Eq for Setup {}
+
+impl std::hash::Hash for Setup {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.addr.hash(state);
+    }
+}
+
+impl Setup {
+    pub fn new(
+        addr: impl std::net::ToSocketAddrs,
+        username: Option<String>,
+    ) -> Result<Self, Error> {
+        let username: Result<String, Error> = username.map(Ok).unwrap_or_else(|| {
+            let user = String::from_utf8(std::process::Command::new("whoami").output()?.stdout)?;
+            let user = user
+                .split_whitespace()
+                .next()
+                .expect("expect newline after whoami output");
+            Ok(user.to_string())
+        });
+        let username = username?;
+
+        Ok(Self {
+            username,
+            addr: addr
+                .to_socket_addrs()?
+                .next()
+                .ok_or_else(|| format_err!("No socket addresses found"))?,
+            key_path: None,
+            setup_fn: None,
+        })
+    }
+
+    pub fn key_path(self, p: impl AsRef<std::path::Path>) -> Self {
+        Self {
+            key_path: Some(p.as_ref().to_path_buf()),
+            ..self
+        }
+    }
+
+    pub fn setup(
+        self,
+        setup: impl Fn(&mut ssh::Session, &slog::Logger) -> Result<(), Error> + Send + Sync + 'static,
+    ) -> Self {
+        Self {
+            setup_fn: Some(Arc::new(setup)),
+            ..self
+        }
+    }
+}
+
+/// Only one machine is supported per instance of this Launcher, further instances of `Setup`
+/// passed to `launch` will
+/// be ignored, since it doesn't make sense to connect to the same machine twice.
+///
+/// The `impl Drop` of this type is a no-op, since Tsunami can't terminate an existing machine.
+#[derive(Default)]
+pub struct Machine {
+    pub log: Option<slog::Logger>,
```

Comment:
It's pretty tempting to move away from `slog` (which requires you to pass the logger around everywhere) and to something like `tracing` (which doesn't). May be worth looking into. The transition is pretty painless since `tracing` has very similar macros.

Reply:
meh, there are some useful features that passing the logger around enables, like here: https://github.com/jonhoo/tsunami/pull/22/files#diff-b4aea3e418ccdb71239b96952d9cddb6R212

Reply:
I'm not sure how that link is relevant?

Reply:
Yeah that link seems wrong. I was referring to adding "default fields" to a logger which are currently used to tag loggers with the region they are associated with, etc.

Reply:
In `tracing` the equivalent is to construct and enter a "`Span`". Converting should probably be done in a separate PR though.